OSPF multi-instance topotest

gromit1811 opened the pull request:
This PR adds a topotest testing OSPF multi-instance support. It also adds/modifies a bit of infrastructure for topotests in general to be able to implement multi-instance topotests at all. And finally, it fixes "show logging" to also include the reporting daemon's instance number and several "show ..." commands to still report non-default daemon instances (ID >= 1) even if the default daemon instance currently isn't running.

4 individual commits for the items above, all in a single PR because they're so tightly coupled. I hope that's OK.

gromit1811 committed:
tests: Add OSPF multi instance topotest

Test multiple OSPF instances running on a router using independent LSDBs but
(unidirectionally) exchanging routes from one instance to the other.

Signed-off-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>

gromit1811 committed:
lib: Make "show logging" also print the daemon instance number

Prefix "show logging" output of each daemon with its instance number (for
non-default instances, i.e. instance ID >= 1).

Useful in general to understand who's reporting what and required by
topotest multi-instance support to determine which instances of which daemon
are actually running.

Signed-off-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/647801926a6f0bf3ccca2e41d5ec2c9a/raw/96b7d89df9da87ba777a027971c94a36e5806472/cr_8126_1614007545.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/topotest.py b/tests/topotests/lib/topotest.py
index 8043bb2f8..ea96b4ff2 100644
--- a/tests/topotests/lib/topotest.py
+++ b/tests/topotests/lib/topotest.py
@@ -1486,12 +1486,15 @@ class Router(Node):
             (daemon_name, daemon_instance) = split_daemon_name_instance(daemon)
             daemon_path = os.path.join(self.daemondir, daemon_name)
             if daemon_instance:
-              instance_args = "-n " + daemon_instance
+                instance_args = "-n " + daemon_instance
             else:
-              instance_args = ""
+                instance_args = ""
             self.cmd(
                 "ASAN_OPTIONS=log_path={3}.asan {0} {1} {2} --log file:{3}.log --log-level debug -d > {3}.out 2> {3}.err".format(
-                    daemon_path, self.daemons_options.get(daemon, ""), instance_args, daemon
+                    daemon_path,
+                    self.daemons_options.get(daemon, ""),
+                    instance_args,
+                    daemon,
                 )
             )
             logger.debug("{}: {} {} started".format(self, self.routertype, daemon))
diff --git a/vtysh/vtysh.c b/vtysh/vtysh.c
index 00845ee06..501cbef55 100644
--- a/vtysh/vtysh.c
+++ b/vtysh/vtysh.c
@@ -461,7 +461,7 @@ static void vtysh_client_config(struct vtysh_client *head_client, char *line)
 
 static bool any_instance_connected(struct vtysh_client *client)
 {
-	for ( ; client; client = client->next) {
+	for (; client; client = client->next) {
 		if (client->fd >= 0)
 			return true;
 	}
@@ -640,7 +640,7 @@ static int vtysh_execute_func(const char *line, int pager)
 				if (vtysh_client[i].fd < 0
 				    && (cmd->daemon == vtysh_client[i].flag)) {
 					if (!any_instance_connected(
-							&vtysh_client[i])) {
+						    &vtysh_client[i])) {
 						fprintf(stderr,
 							"%s is not running\n",
 							vtysh_client[i].name);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

gromit1811 committed:
vtysh: Fix multi-instance support for various "show ..." commands

This extends a4e2dac168da8c3884288f023451308f0bacb726 to other cases where
multi-instance daemons should be checked for explicitly. It also puts the
actual "daemon connected" check into a separate function to make the purpose
of the check easier to understand.

Signed-off-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>

gromit1811 committed:
tests: Topotest mult-instance support

General topotest infrastructure and OSPF-specific multi-instance features.
Supports max. 2 instances in addition to the default instance.

Signed-off-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>

odd22 reviewed (approved):
LGTM. Need to rebase the code.

I would also suggest to use the json output of the routing table instead of text to assess the test.

gromit1811 commented:
> LGTM. Need to rebase the code.

Thanks for the review. I'll take care of rebasing - there was an independent fix for the same problem in vtysh implemented in a slightly different way causing the conflict.

> I would also suggest to use the json output of the routing table instead of text to assess the test.

Do you mean the convergence check in test_ospf_converged() ("show ip ospf X neigh") or the routing table checks in test_ospf_routingTable()/test_ospf_kernel_route()? The former is mostly copy/paste from ospf6-topo1, but I can give the JSON variant a try. The latter two use topotest infrastructure which I wouldn't want to touch in this PR.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/lib/topotest.py:
************* Module lib.topotest
my_frr/tests/topotests/lib/topotest.py:1426:16: E1305: Too many arguments for format string (too-many-format-args)
my_frr/tests/topotests/lib/topotest.py:1441:16: E1305: Too many arguments for format string (too-many-format-args)
my_frr/tests/topotests/lib/topotest.py:1834:4: E0213: Method should have "self" as first argument (no-self-argument)
my_frr/tests/topotests/lib/topotest.py:1835:24: E0602: Undefined variable 'Self' (undefined-variable)
my_frr/tests/topotests/lib/topotest.py:1851:15: E0602: Undefined variable 'unicode' (undefined-variable)

-----------------------------------
Your code has been rated at 9.73/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

gromit1811 commented:
> I would also suggest to use the json output of the routing table instead of text to assess the test.

What do you think about this version of `test_ospf_converged()`? This now uses the JSON variant of "show ip ospf neighbor". IMO, the code is a bit easier to read (but mostly because I'm not setting/resetting `isConverged` all the time). OTOH, if you're used to the normal output of the neighbor command and are not familiar with the JSON one, it could be more difficult.

```
def test_ospf_converged():
    "Check whether all routers have converged"

    tgen = get_topogen()

    # Wait for OSPF to converge (All Neighbors in Full State)
    logger.info("Waiting for OSPF convergence")

    res = "NONE"
    timeout = 60
    while timeout > 0:
        logger.info("Timeout in %s: " % timeout),
        sys.stdout.flush()

        # Look for any node not yet converged
        for router, rnode in tgen.routers().items():
            isConverged = False
            for daemon, started in sorted(rnode.net[router].daemons.items()):
                if not daemon.startswith("ospfd") or not started:
                    continue

                (d, instance) = topotest.split_daemon_name_instance(daemon)
                instanceArg = " {}".format(instance) if instance else ""

                res = rnode.vtysh_cmd("show ip ospf" + instanceArg
                                      + " neighbor json", isjson=True)

                neighData = chain.from_iterable(
                    res.get("neighbors", {}).values()
                )
                convStates = [n.get("state", "").startswith("Full")
                              for n in neighData]
                isConverged = len(convStates) and all(convStates)

                if not isConverged:
                    logger.info("Waiting for {}{}".format(router, instanceArg))
                    sys.stdout.flush()
                    break

            if not isConverged:
                break

        if isConverged:
            logger.info("Done")
            break
        else:
            sleep(5)
            timeout -= 5

    if timeout <= 0:
        # Bail out with error if a router fails to converge
        assert False, "OSPF did not converge:\n{}".format(res)

    logger.info("OSPF converged.")

    # For debugging after OSPF convergence
    if os.environ.get("TOPOTESTS_MININET_CLI") == "ospf-multi-instance:converged":
        tgen.mininet_cli()

    # Make sure that all daemons are still running
    if tgen.routers_have_failure():
        assert tgen.errors == "", tgen.errors
```

Should I push this one instead?

As mentioned above, I've not touched the actual routing table checks to avoid having to modify the topotest infrastructure.

gromit1811 commented:
@odd22 What do you think about the changes above using the JSON output for the convergence check? Should I push that one instead of the original version? As I wrote there, I think the new version has both advantages and disadvantages and I can perfectly live with both variants, so it depends on you which one you prefer.

gromit1811 commented:
Rebased to current master

@odd22 @donaldsharp @mjstapp Ping :-)
I assume your even more busy than I am, but could you spare some time to have a look at this? I've got another PR in the pipeline partially dependent on this one. Thanks!

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/lib/topotest.py:
************* Module lib.topotest
my_frr/tests/topotests/lib/topotest.py:1934:15: E0602: Undefined variable 'unicode' (undefined-variable)

-----------------------------------
Your code has been rated at 9.95/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

gromit1811 commented:
Added pytest mark, rebased to current master

odd22 commented:
LGTM

odd22 commented:
ci:rerun

odd22 commented:
It seems that his PR introduces a problem with other topotests. Can you check and add appropriate correction ?

gromit1811 commented:
> It seems that his PR introduces a problem with other topotests. Can you check and add appropriate correction ?

I'd love to, but I've got absolutely no idea about how my change could possibly cause a failure in a BGP topotest. I'm neither touching BGP nor BGP topotest code. And the tiny bit of common code in lib and vtysh I'm modifying IMO is very unlikely to cause such failures, especially when the very same code seems to work fine on different platforms.

I'd rather assume that this is a problem in the BGP topotests causing intermittent failures. It seems there was an attempt recently to fix such failures in 2268cf50d596c854e9b50316ddc33e2b7d5867d9 but that commit is present in my PR and doesn't seem to help.
@kuldeepkash: Could it be that your commit doesn't fix all instances of this problem or are we looking at something completely different here?

kuldeepkash commented:
> > It seems that his PR introduces a problem with other topotests. Can you check and add appropriate correction ?
> 
> I'd love to, but I've got absolutely no idea about how my change could possibly cause a failure in a BGP topotest. I'm neither touching BGP nor BGP topotest code. And the tiny bit of common code in lib and vtysh I'm modifying IMO is very unlikely to cause such failures, especially when the very same code seems to work fine on different platforms.
> 
> I'd rather assume that this is a problem in the BGP topotests causing intermittent failures. It seems there was an attempt recently to fix such failures in [2268cf5](https://github.com/FRRouting/frr/commit/2268cf50d596c854e9b50316ddc33e2b7d5867d9) but that commit is present in my PR and doesn't seem to help.
> @kuldeepkash: Could it be that your commit doesn't fix all instances of this problem or are we looking at something completely different here?

Hi @gromit1811, I am looking into BGP-GR test failure, I have raised a PR: #8612 for the same.

ton31337 reviewed (changes requested):
Please rename tests/topotest directories to follow new guidelines:

```
Directory name for a new topotest must not contain hyphen (``-``) characters.
  To separate words, use underscores (``_``). For example, `tests/topotests/bgp_new_example`.
```

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


---

Pylint found errors in source files changed by this PR:

```
Pylint report for my_frr/tests/topotests/lib/topotest.py:
************* Module lib.topotest
my_frr/tests/topotests/lib/topotest.py:1967:15: E0602: Undefined variable 'unicode' (undefined-variable)

-----------------------------------
Your code has been rated at 9.95/10




```


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

gromit1811 commented:
> Please rename tests/topotest directories to follow new guidelines:

Done
